services/token_airdrop.proto

```diff
@@ -0,0 +1,104 @@
+// <<<pbj.java_package = "com.hedera.hapi.node.token">>> This comment is special code for setting PBJ Compiler java package
+option java_multiple_files = true;
+
+import "basic_types.proto";
+
+/**
+ * Airdrops one or more tokens to one or more accounts.
+ *
+ * ### Effects
+ * This distributes tokens from the balance of one or more sending account(s) to the balance
+ * of one or more recipient accounts. Accounts MAY receive the tokens in one of four ways.
+ *
+ *  - An account already associated to the token to be distributed SHALL receive the
+ *    airdropped tokens immediately to the recipient account balance.<br/>
+ *    The fee for this transfer SHALL include the transfer, the airdrop fee, and any custom fees.
+ *  - An account with available automatic association slots SHALL be automatically
+ *    associated to the token, and SHALL immediately receive the airdropped tokens to the
+ *    recipient account balance.<br/>
+ *    The fee for this transfer SHALL include the transfer, the association, the cost to renew
+ *    that association once, the airdrop fee, and any custom fees.
+ *  - An account with "receiver signature required" set SHALL have a "Pending Airdrop" created
+ *    and must claim that airdrop with a `claimAirdrop` transaction.<br/>
+ *    The fee for this transfer SHALL include the transfer, the association, the cost to renew
+ *    that association once, the airdrop fee, and any custom fees. If the pending airdrop is not
+ *    claimed immediately, the `sender` SHALL pay the cost to renew the token association, and
+ *    the cost to maintain the pending airdrop, until the pending airdrop is claimed or cancelled.
+ *  - An account with no available automatic association slots SHALL have a "Pending Airdrop"
+ *    created and must claim that airdrop with a `claimAirdrop` transaction.<br/>
+ *    The fee for this transfer SHALL include the transfer, the association, the cost to renew
+ *    that association once, the airdrop fee, and any custom fees. If the pending airdrop is not
+ *    claimed immediately, the `sender` SHALL pay the cost to renew the token association, and
+ *    the cost to maintain the pending airdrop, until the pending airdrop is claimed or cancelled.
+ *
+ * If an airdrop would create a pending airdrop for a fungible/common token, and a pending airdrop
+ * for the same sender, receiver, and token already exists, the existing pending airdrop
+ * SHALL be updated to add the new amount to the existing airdrop, rather than creating a new
+ * pending airdrop.
+ *
+ * Any airdrop that completes immediately SHALL be irreversible. Any airdrop that results in a
+ * "Pending Airdrop" MAY be canceled via a `cancelAirdrop` transaction.
+ *
+ * All transfer fees (including custom fees and royalties), as well as the rent cost for the
+ * first auto-renewal period for any automatic-association slot occupied by the airdropped
+ * tokens, SHALL be charged to the account paying for this transaction.
```

Comment:
It isn't clear how the automatic-association slot and renewal fees will be charged
to the payer currently.  Something to bring up in the design discussions, perhaps.

Reply:
If the recipient has available auto association slots, they will be used directly during an airdrop. In this case the payer of the airdrop will pay:
- the usage of this auto association
- the first rental auto-renew period for the token

In this case the token will be directly transferred.

If the recipient has no associations with token and no available auto association slots, then the payer of the airdrop will pay:
- prepayment for the association
- the first rental auto-renew period for the token
- the fee for transfer operation
- the airdrop fee itself
- all custom fees required to be paid by the sender/payer

In this case the token will end up in pending state.

Reply:
I am aware of how the association is intended.  What I meant is that we have no mechanism, currently, to charge the payer for the renewal fee, as we cannot "store" fees, and we have nowhere in state to record that a renewal fee is "pre-paid".
I think that there is a design gap here, and I do not believe that we _can_ do this under current design and structure.